src/sink/kafka.rs

```diff
@@ -0,0 +1,214 @@
+    ) {
+        let key = format!("{:X}", order_by);
+        let future = self.try_payload(bytes.as_slice(), key.as_bytes());
+        self.messages.push(future);
+    }
+
+    fn flush(&mut self) {
+        let retry_payload_and_keys: Vec<Option<OwnedMessage>> = self.messages
+            .iter_mut()
+            .map(|future| {
+                let result = future.wait();
+                match result {
+                    Ok(inner) => match inner {
+                        Ok((_partition, _offset)) => {
+                            KAFKA_PUBLISH_SUCCESS_SUM.fetch_add(1, Ordering::Relaxed);
+                            None
+                        }
+
+                        Err((err, message)) => match err {
+                            KafkaError::MessageProduction(err) => match err {
+                                RDKafkaError::InvalidMessage
+                                | RDKafkaError::UnknownTopicOrPartition
+                                | RDKafkaError::LeaderNotAvailable
+                                | RDKafkaError::NotLeaderForPartition
+                                | RDKafkaError::RequestTimedOut
+                                | RDKafkaError::NetworkException
+                                | RDKafkaError::GroupLoadInProgress
+                                | RDKafkaError::GroupCoordinatorNotAvailable
+                                | RDKafkaError::NotCoordinatorForGroup
+                                | RDKafkaError::NotEnoughReplicas
+                                | RDKafkaError::NotEnoughReplicasAfterAppend
+                                | RDKafkaError::NotController => {
+                                    KAFKA_PUBLISH_RETRY_SUM
+                                        .fetch_add(1, Ordering::Relaxed);
+                                    Some(message)
+                                }
+
+                                _ => {
+                                    error!("Kafka broker returned an unrecoverable error: {:?}", err);
+                                    KAFKA_PUBLISH_FAILURE_SUM
+                                        .fetch_add(1, Ordering::Relaxed);
+                                    None
+                                }
+                            },
+
+                            _ => {
+                                error!("Failed in send to kafka broker: {:?}", err);
+                                KAFKA_PUBLISH_FAILURE_SUM
+                                    .fetch_add(1, Ordering::Relaxed);
+                                None
+                            }
+                        },
+                    },
+
+                    _ => {
+                        error!("Failed in send to kafka broker: {:?}", result);
+                        KAFKA_PUBLISH_FAILURE_SUM.fetch_add(1, Ordering::Relaxed);
+                        None
+                    }
+                }
+            })
+            .collect();
+        let new_messages = retry_payload_and_keys
```

Comment:
Your rust-fu may be stronger than my own.

Where do we retry events that have failed republishing multiple times?